fdb-relational-core/src/main/java/com/apple/foundationdb/relational/recordlayer/query/QueryParser.java

```diff
@@ -156,6 +157,87 @@ public static RelationalParser.SqlInvokedFunctionContext parseFunction(@Nonnull
         return result;
     }
 
+    @SuppressWarnings("PMD.AvoidStringBufferField") // the lifetime of this object is very short, and it is scoped to processing DDLs of temp functions only.
+    private static final class PreparedParamsReplacer extends RelationalParserBaseVisitor<Void> {
+
+        @Nonnull
+        private final StringBuilder unpreparedSqlBuilder;
+
+        @Nonnull
+        private final StringBuilder sqlBuilder;
+
+        @Nonnull
+        private final PreparedParams preparedStatementParameters;
+
+        private PreparedParamsReplacer(@Nonnull final PreparedParams preparedStatementParameters) {
+            this.preparedStatementParameters = preparedStatementParameters;
+            unpreparedSqlBuilder = new StringBuilder();
+            sqlBuilder = new StringBuilder();
+        }
+
+        @Override
+        public Void visitPreparedStatementParameter(@Nonnull RelationalParser.PreparedStatementParameterContext ctx) {
+            Object param;
+            if (ctx.QUESTION() != null) {
+                param = preparedStatementParameters.nextUnnamedParamValue();
+            } else {
+                final var namedParameterContext = ctx.NAMED_PARAMETER();
+                final var parameterName = namedParameterContext.getText().substring(1);
+                param = preparedStatementParameters.namedParamValue(parameterName);
+            }
+            unpreparedSqlBuilder.append(PreparedParams.prettyPrintParam(param)).append(" ");
+            return null;
+        }
+
+        @Override
+        public Void visitTerminal(@Nonnull TerminalNode node) {
+            if (node.getSymbol().getType() != Token.EOF) {
+                unpreparedSqlBuilder.append(node.getText()).append(" ");
+                sqlBuilder.append(node.getText()).append(" ");
+            }
+            return null;
+        }
+
+        @Nonnull
+        String getQuery() {
+            return unpreparedSqlBuilder.toString();
+        }
+    }
+
+    private static final class PreparedParamsValidator extends RelationalParserBaseVisitor<Void> {
+        @Override
+        public Void visitPreparedStatementParameter(final RelationalParser.PreparedStatementParameterContext ctx) {
+            Assert.failUnchecked(ErrorCode.SYNTAX_ERROR, "found prepared parameter(s) in SQL statement");
+            return null;
+        }
+    }
+
+    /**
+     * Replaces the prepared parameters with their corresponding literals, note that the resulting query is not necessarily
```

Comment:
I am not sure I fully understand this. 
Does this mean we encode the value of the literals into what ultimately becomes the cache key? 
If so, I don't think that this is what we would want, since we would need different values for the same parameter to still result in the same cache entry, similar to how we handle queries.